### ISSUE TITLE
Cannot interrupt service server using fastrtps or connext

### ISSUE THREAD

**Issue by sloretz** [ros2/rclpy#195]
## Bug report

**Required Info:**

- Operating System:
  - Windows 10 10.0.10240
- Installation type:
  - binaries
- Version or commit hash:
  - https://ci.ros2.org/view/packaging/job/packaging_windows/1046/
- DDS implementation:
  - RTI Connext, Fast-RTPS
- Client library (if applicable):
  - rclpy

#### Steps to reproduce issue

Start `add_two_ints_server.exe` using either fastrtps or connext

```bat
set RMW_IMPLEMENTATION=rmw_fastrtps_cpp
C:\dev\ros2-windows\Lib\demo_nodes_py\add_two_ints_server.exe
```

```bat
set RMW_IMPLEMENTATION=rmw_connext_cpp
C:\dev\ros2-windows\Lib\demo_nodes_py\add_two_ints_server.exe
```

Try to interrupt the server using `CTRL` + `c`

#### Expected behavior

Processes would print a traceback ending in `KeyboardInterrupt` and exit

#### Actual behavior

No apparent response. Process does not print anything to console, and must be killed using the task manager.

#### Additional information
I see the expected behavior using `rmw_opensplice_cpp`

**Comment by sloretz** [comment]
Confirmed that `demo_nodes_cpp` service server can be interrupted no problem, so this is just an rclpy issue

**Comment by dirk-thomas** [comment]
I just tried it with sourcing the referenced build on the same Windows machine (windshield) and it was interrupt able without a problem (just tried with the default FastRTPS):

```
C:\J\workspace>packaging_windows\ws\install\Lib\demo_nodes_py\add_two_ints_server.exe
Traceback (most recent call last):
  File "C:\J\workspace\packaging_windows\ws\install\Lib\demo_nodes_py\add_two_ints_server-script.py", line 11, in <module>
    load_entry_point('demo-nodes-py==0.4.0', 'console_scripts', 'add_two_ints_server')()
  File "C:\J\workspace\packaging_windows\ws\install\Lib\site-packages\demo_nodes_py\services\add_two_ints_server.py", line 39, in main
    rclpy.spin(node)
  File "C:\J\workspace\packaging_windows\ws\install\Lib\site-packages\rclpy\__init__.py", line 108, in spin
    executor.spin_once()
  File "C:\J\workspace\packaging_windows\ws\install\Lib\site-packages\rclpy\executors.py", line 519, in spin_once
    handler, entity, node = self.wait_for_ready_callbacks(timeout_sec=timeout_sec)
  File "C:\J\workspace\packaging_windows\ws\install\Lib\site-packages\rclpy\executors.py", line 505, in wait_for_ready_callbacks
    return next(self._cb_iter)
  File "C:\J\workspace\packaging_windows\ws\install\Lib\site-packages\rclpy\executors.py", line 411, in _wait_for_ready_callbacks
    _rclpy.rclpy_wait(wait_set, timeout_nsec)
KeyboardInterrupt
```

Same result with the last nightly debug buiild on portable.

**Comment by sloretz** [comment]
The windows version between my VM and the build machine are different (10.0.10240 vs 10.0.17324.112). I'm struggling to find any other difference. If a third machine doesn't show the same issue I would vote to close this.

**Comment by dirk-thomas** [comment]
On icecube a build from a couple of days works too.

**Comment by sloretz** [comment]
Retested using https://ci.ros2.org/view/packaging/job/packaging_windows/1060/ on my VM and on windows_portable. The behavior has changed :(

Using my VM there is no response to `CTRL` + `c` using fastrtps and connext,  and now there is also no response when using opensplice. I'm still able to interrupt the `demo_nodes_cpp` server no problem. 

Using windows_portable the first time I tried to `CTRL` + `c` `demo_nodes_py/add_two_ints_server.exe` it hung. I tried to interrupt it multiple times over a few minutes and it never responded. I opened another cmd window and every time I am able to interrupt it using any rmw implementation.

Maybe there is a race condition that is made worse on earlier windows 10 versions? Portable is Windows 10 Enterprise v1803 while my VM is enterprise v10240 (older despite being a bigger number)
